external_services/lti.py

```diff
@@ -126,26 +136,32 @@ def sign_get_query(self, url=None):
 
 class CustomStudentInfoLTIRequest(LTIRequest):
 
-    def __init__(self, service, user, profiles, instance, host, title, context_id=None, link_id=None, add=None):
+    def __init__(self, service, user, profiles, instance, request, title, context_id=None, link_id=None, add=None, exercise=None):
+        self.service = service
+        self.course_instance = instance
         parameters = add or {}
-        parameters['custom_student_id'] = self.true_student_id(user.userprofile)
+        if not service.is_anonymous:
+            parameters['custom_student_id'] = self.true_student_id(user.userprofile)
         if len(profiles) > 1:
             parameters['custom_group_members'] = self.group_json(profiles)
-        super().__init__(service, user, instance, host, title, context_id, link_id, parameters)
+        super().__init__(service, user, instance, request, title, context_id, link_id, parameters, exercise)
 
     def true_student_id(self, profile):
         return profile.student_id or "A{:d}".format(profile.id)
 
     def group_json(self, profiles):
-        data = [];
+        data = []
         for profile in profiles:
             user = profile.user
```

Comment:
This could be moved to next line.

Reply:
Where? You can not initialize the list inside the loop.

Reply:
```
user = profile.user
student_id, given_name, family_name, full_name, email = self.user_info(self.course_instance, user)
```

to

```
student_id, given_name, family_name, full_name, email = self.user_info(self.course_instance, profile.user)
```